proposals/evm_rpc_2024_10_18.md

```diff
@@ -0,0 +1,163 @@
+# Proposal to reinstall the EVM RPC canister
+
+Repository: `https://github.com/internet-computer-protocol/evm-rpc-canister.git`
+
+Git hash: `3de8830e509cb97411367a153494e524573903f4`
+
+New compressed Wasm hash: `636442e4349a007316dcc1452d9a256d22203186181d2534f33bd264162305b3`
+
+Target canister: `7hfb6-caaaa-aaaar-qadga-cai`
+
+Previous EVM RPC proposal: https://dashboard.internetcomputer.org/proposal/132550
+
+---
+
+## Motivation
+
+This proposal does a complete re-installation of the EVM RPC canister using [v2.2.0](https://github.com/internet-computer-protocol/evm-rpc-canister/releases/tag/v2.2.0).
+The main changes are:
+* Immutable providers: providers are now hard-coded and could only be changed via a NNS proposal. 
+    * The URLs used are hard-coded and contain a placeholder for an API key, in case the provider does not support bearer authentication which is otherwise preferred.
+    * Privileged principals (see install args below) can manage the API keys at runtime, without requiring a NNS proposal. The provided keys are validated by the canister, ensuring for example that URL escaping is not possible. Excepted for API keys, those privileged principals cannot change in any way the specified providers.
+    * See this [forum post](https://forum.dfinity.org/t/evm-rpc-canister/23313/53) for more information.
+* Support a new threshold strategy to aggregate responses from multiple providers. This allows callers to be more robust against a single provider failure.
+* Added a new convenience method for `eth_call` to allow calling read-only functions of a smart contract.
+
+## Install args
+
+```
+git fetch
+git checkout 3de8830e509cb97411367a153494e524573903f4
+didc encode -d evm_rpc.did -t '(InstallArgs)' '(record {demo = opt false; manageApiKeys = opt vec {principal "mf7xa-laaaa-aaaar-qaaaa-cai"}; logFilter = null})' | xxd -r -p | sha256sum
```

Comment:
`4449444c066c03dcc690fd0301a3b299930403c98fb9b104046e026b0484809ec7047fdf92a2a4087fb3f594920d718e9bd6980f716e7e6e056d6801000001000101010a00000000023000000101`

Reply:
4449444c066c03dcc690fd0301a3b299930403c98fb9b104046e026b0484809ec7047fdf92a2a4087fb3f594920d718e9bd6980f716e7e6e056d6801000001000101010a00000000023000000101

But there are two problems:
1. it should be `candid/evm_rpc.did`.
2. The above hex is the output before `xxd` command. So line 31 is also doing sha256sum, not sure if this is what you want.

Reply:
Thanks @ninegua for checking!
> it should be candid/evm_rpc.did.

Absolutely right, good catch!

> The above hex is the output before xxd command. So line 31 is also doing sha256sum, not sure if this is what you want.

Actually yes, with the new proposal types only the hash of the args is displayed, see this [thread](https://dfinity.slack.com/archives/C0175CHJ0P6/p1723237367803209)

Reply:
I don't get the same hash. I also stumbled over candid/evm_rpc.did. After fixing this, I get

```
$ didc encode -d candid/evm_rpc.did -t '(InstallArgs)' '(record {demo = opt false; manageApiKeys = opt vec {principal "mf7xa-laaaa-aaaar-qaaaa-cai"}; logFilter = null})' | xxd -r -p | sha256sum
3d28b9b62643bc7875c5c611b13bdfb39aef6f342fa8af74ad0a5e8998003f31

$ git status
HEAD detached at 3de8830
```

Reply:
@THLO Sorry for the confusion,

> 3d28b9b62643bc7875c5c611b13bdfb39aef6f342fa8af74ad0a5e8998003f31

is the correct hash of

> 4449444c066c03dcc690fd0301a3b299930403c98fb9b104046e026b0484809ec7047fdf92a2a4087fb3f594920d718e9bd6980f716e7e6e056d6801000001000101010a00000000023000000101

so everything is good ✅ . To double check you can do

```
echo "4449444c066c03dcc690fd0301a3b299930403c98fb9b104046e026b0484809ec7047fdf92a2a4087fb3f594920d718e9bd6980f716e7e6e056d6801000001000101010a00000000023000000101" | xxd -r -p | sha256sum
```